### PR TITLE
Added Gaussian Process Support via GPy

### PR DESCRIPTION
**Pull request by cxhernandez**
I'm adding a new Gaussian process strategy using [GPy](https://sheffieldml.github.io/GPy/). It seemed to work pretty well in Osprey's toy example case, converging in <10 iterations.

I also removed MOE as a search strategy, as it seems to no longer be supported and was not very effective (re https://github.com/pandegroup/osprey/issues/96).

### CODE REVIEW

**Comment by rmcgibbo** [comment]
Cool!

**Comment by cxhernandez** [comment]
I'll try to build a conda package for GPy at some point today....

File: osprey/strategies.py [hunk 1]
```diff
@@ -168,61 +172,45 @@ def _hyperopt_fmin_random_kwarg(random):
...
+        if not init:
+            init = self._init()
+
+        def z(x):
+            y = x.copy().reshape(-1, self.n_dims)
+            s, v = self.model.predict(y)
+            s[(x < 0.)*(x > 1.)] = -1
+            v[(x < 0.)*(x > 1.)] = -1
+            if uncertainty:
+                return -v.flatten()
```
Comment (rmcgibbo) [inline osprey/strategies.py]: Is this right? I dont know the api, but don't we want to add the mean and std dev or something to minimize the upper edge of the confidence interval or something?

Comment (cxhernandez) [inline osprey/strategies.py]: Here, `s` are the predicted mean scores and `v` are the variances about the mean.

This `uncertainty` flag is just to change up the strategy from searching for the optimal score to searching in a region where we have less certainty in our prediction. I guess it's kinda hack-y.

So you think something like `return -(s + v).flatten()` would be best?

**Comment by rmcgibbo** [comment]
Let's talk in lab.

**Comment by cxhernandez** [comment]
heading to BPS today
:airplane: :arrow_right: :city_sunset: 

skype?

**Comment by rmcgibbo** [comment]
I'll just type when I get a sec.

On Fri, Feb 26, 2016 at 9:18 AM, Carlos Hernández <notifications@github.com>
wrote:

> heading to BPS today
> [image: :airplane:] [image: :arrow_right:] [image: :city_sunset:]
>
> skype?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pandegroup/osprey/pull/97#issuecomment-189379086>.
>



-- 
-Robert

**Comment by cxhernandez** [comment]
@rmcgibbo: is this better?

**Comment by cxhernandez** [comment]
Okay. Turns out GPy doesn't support py2.6, so I had to exclude it from that particular Travis build.

Does anyone else want to give this a try and see if it works for them (or have any suggestions)?

cc @msultan @mpharrigan @evanfeinberg

**Comment by mpharrigan** [comment]
You should drop 2.6 from Travis

**Comment by cxhernandez** [comment]
@msultan suggested we try to incorporate the test score variances into the model as well.

**Comment by cxhernandez** [comment]
Okay, I changed how `history` is generated so that it includes all CV scores, not just mean. This lets us include the observed model variance into the GP, which improves our oracle's predictions.

**Comment by cxhernandez** [comment]
For a real data set with 4 tunable parameters, I was able to match random search in both GMRQ and parameter values in about half the number of iterations (600 vs. 1200).

It'd be nice if there was a way to reduce the search window after some amount of certainty is establish, so the GP optimization isn't initialized with just some [random point](https://github.com/pandegroup/osprey/pull/97/files#diff-73743f4bcdb2c873cf206b6e5ccc20a3R202).

**Pull request merged**